src/MigrationTask.js

```diff
@@ -1,75 +1,51 @@
+const MigrationError = require('./error/MigrationError');
 const Task = require('./Task');
 
+/**
+ * @type {module.MigrationTask}
+ */
+
 module.exports = class MigrationTask extends Task {
 
-    constructor(task, rerun = false) {
-        const name = `Migration:${task.name}`;
-        super(name, task.version);
+    constructor(task) {
+        const identifier = `Migration:${task.identifier}`;
+        const { version, options } = task;
+        super({identifier, version, options});
+
         this.task = task;
-        this.rerun = rerun;
     }
 
-    async run(dependencies) {
+    /**
+     * @param app
+     * @return {Promise.<*>}
+     */
+    async run({ MigrationModel, transaction }) {
 
-        await this.ensureDependencies(dependencies, ['app']);
+        const ranSuccessfully = await MigrationModel.taskRanSuccessfully(this.task, transaction);
 
-        const migrationModel = dependencies.app.models.Migration;
-        const migrationRanBefore = await this.migrationAlreadyPerformed(migrationModel);
-
-        if (migrationRanBefore && this.rerun === false) return Promise.resolve();
+        if (ranSuccessfully) {
+            return null;
+        }
 
-        const migration = await this.createMigration(migrationModel);
+        const migration = await MigrationModel.startTask(this.task, transaction);
 
         try {
-            const result = await this.runTask(dependencies);
-            return this.finalizeMigration(migration, result);
+            const result = await this.task.run({ MigrationModel, transaction });
+            return this.finalizeMigration(migration, result, null, transaction);
```

Comment:
Not sure about this, I am not catching errors here to make sure it is not called twice